ros2/src/autorccar_costmap/src/costmap.cpp

```diff
@@ -0,0 +1,227 @@
+#include "costmap.h"
+
+#include <yaml-cpp/yaml.h>
+
+#include <fstream>
+#include <iostream>
+
+Costmap::Costmap(const std::string& config_file_path) : costmap_flag(false), cnt_iter(0) {
+    LoadConfig(config_file_path);
+
+    // Initialize global costmap
+    global_size_x = (int)global_width / resolution;
+    global_size_y = (int)global_height / resolution;
+    global_center_x = global_size_x / 2;
+    global_center_y = global_size_y / 2;
+    InitCostmap(global_costmap, global_size_x, global_size_y);
+
+    // Initialize local costmap
+    local_size_x = (int)local_width / resolution;
+    local_size_y = (int)local_height / resolution;
+    local_center_x = local_size_x / 2;
+    local_center_y = local_size_y / 2;
+    InitCostmap(local_costmap, local_size_x, local_size_y);
+
+    // Allocate memory
+    cur_point_cloud.reset(new pcl::PointCloud<pcl::PointXYZI>());
+    cur_point_cloud_transformed.reset(new pcl::PointCloud<pcl::PointXYZI>());
+    cur_point_cloud_filtered.reset(new pcl::PointCloud<pcl::PointXYZI>());
+}
+
+void Costmap::LoadConfig(const std::string& config) {
+    std::ifstream fin(config);
+    if (!fin) {
+        std::cerr << "Failed to open file: " << config << std::endl;
+        exit(-1);
+    }
+    fin.close();
+
+    YAML::Node cfg = YAML::LoadFile(config);
+
+    resolution = cfg["global"]["resolution"].as<double>();
+    cnt_limit = cfg["global"]["updateEveryNthLidar"].as<unsigned int>();
+
+    global_width = cfg["global"]["width"].as<int>();
+    global_height = cfg["global"]["height"].as<int>();
+
+    local_width = cfg["local"]["width"].as<int>();
+    local_height = cfg["local"]["height"].as<int>();
+
+    std::cout << "Successfully loaded the config" << std::endl;
+}
+
+void Costmap::InitCostmap(Eigen::MatrixXd& costmap, int size_x, int size_y) {
+    costmap.resize(size_x, size_y);
+    costmap.setConstant(ProbabilityToLogOdds(P_UNKNOWN));
+}
+
+void Costmap::UpdatePointCloud(const pcl::PointCloud<pcl::PointXYZI>::Ptr pcl) {
+    cur_point_cloud->clear();
+    cur_point_cloud_transformed->clear();
+    cur_point_cloud_filtered->clear();
+
+    pcl::copyPointCloud(*pcl, *cur_point_cloud);
+
+    UpdateCostmapFlag();
+}
+
+void Costmap::UpdatePose(Eigen::Matrix<double, 4, 4>& trans) { cur_pose = trans; }
+
+void Costmap::UpdateCostmapFlag() {
+    cnt_iter += 1;
+
+    if (cnt_iter >= cnt_limit) {
+        costmap_flag = true;
+        cnt_iter = 0;
+    }
+}
+
+void Costmap::UpdateCostmap() {
+    // Get cell indices of current pose. This will be used as the start point of the bresenham line.
+    int sx = floor(cur_pose(0, 3) / resolution) + global_center_x;  // assume (global_center_x, global_center_y) = (0,0)
+    int sy = floor(cur_pose(1, 3) / resolution) + global_center_y;
```

Comment:
매우 강력 권장)
```suggestion
    const int sx = floor(cur_pose(0, 3) / resolution) + global_center_x;  // assume (global_center_x, global_center_y) = (0,0)
    const int sy = floor(cur_pose(1, 3) / resolution) + global_center_y;
```

**Add const as possible**

당최 원인을 알 수 없는 버그를 몇시간째 고치느라 그 어떤것도 믿을 수 없는 의심암귀의 구렁텅이에 빠졌을때
`const`가 여기저기 붙어있다는게 조금 도움이 될 때가 있습니다.

그리고 컴파일러에게 const 변수라는것을 알려줘서 최적화에 도움이 될 수 있다고도 들었습니다.